Prepare the way for incremental repair

raphaelsc opened the pull request:
With incremental repair, each replica::compaction_group will have 3 logical compaction groups, repaired, repairing and unrepaired. The definition of group is a set of sstables that can be compacted together. The logical groups will share the same instance of sstable_set, but each will have its own logical sstable set. Existing compaction::table_state is a view for a logical compaction group. So it makes sense that each replica::compaction_group will have multiple views. Each view will provide to compaction layer only the sstables that belong to it. That way, we preserve the existing interface between replica and compaction layer, where each compaction::table_state represents a single logical group.
The idea is that all the incremental repair knowledge is confined to repair and replica layer, compaction doesn't want to know about it, it just works on logical groups, what each represents doesn't matter from the perspective of the subsystem. This is the best way forward to not violate layers and reduce the maintenance burden in the long run.
We also proceed to rename table_state to compaction_group_view, since it's a better description. Working with multiple terms is confusing. The placeholder for implementing the sstable classifier is also left in tablet_storage_group_manager, by the time being, all sstables will go to the unrepaired logical set, which preserves the current behavior.

avikivity commented on replica/compaction_group.hh:
```diff
@@ -101,6 +101,8 @@ public:
         return _group_id;
     }
 
```
>replica: Allow compaction_group to have more than one view
>
> In order to support incremental repair, we'll allow each
replica::compaction_group to have two logical compaction groups
(or logical sstable sets), one for repaired, another for unrepaired.

We'll need three, no? repaired, undergoing repair, unrepaired.

avikivity commented on compaction/compaction_group_view.hh:
```diff
@@ -38,8 +38,8 @@ public:
     // min threshold as defined by table.
     virtual unsigned min_compaction_threshold() const noexcept = 0;
     virtual bool compaction_enforce_min_threshold() const noexcept = 0;
```
> replica: Futurize retrieval of sstable sets in compaction_group_view
>
> This will allow upcoming work to gently produce a sstable set for
each compaction group view. Example: repaired and unrepaired.

Not sure it's needed. With tablets, we have a relatively small number of sstables per tablet.

avikivity replied:
Looks like the cause for this is the fact that you don't have separate sstable sets for repaired/unrepaired.

If the only user is compaction/repair, it's not too bad. Compaction/repair also insert sstables, which is a slow/futurized process. But the data path should not use these. Does it?

avikivity replied:
I guess it doesn't, since this is for compaction_group_view.

raphaelsc commented on compaction/compaction_manager.cc:
```diff
@@ -2197,7 +2197,6 @@ future<compaction_manager::compaction_stats_opt> compaction_manager::perform_sst
 
 compaction::compaction_state::compaction_state(compaction_group_view& t)
     : gate(format("compaction_state for table {}.{}", t.schema()->ks_name(), t.schema()->cf_name()))
```
> > Move backlog tracker to replica::compaction_group
> > Since there will be only one physical sstable set, it makes sense to move
> > backlog tracker to replica::compaction_group. With incremental repair,
> > it still makes sense to compute backlog accounting both logical sets,
> > since the compound backlog influences the overall read amplification,
> > and the total backlog across repaired and unrepaired sets can help
> > driving decisions like giving up on incremental repair when unrepaired
> > set is almost as large as the repaired set, causing an amplification
> > of 2.
> 
> But aren't there really two sstable sets?

One instance of sstable_set, two logical sstable sets.

> 
> Since the two sets can't be compacted with each other, raising shares doesn't help anything. Or maybe it does, since we do plan to merge the sets together later. But it can happen it a really long time, and meanwhile, we're raising shares without the extra shares helping to reduce the backlog.
> 
> (the controller logic is backlog grows -> shares grow -> backlog drops, but here it can't happen)

I agree the other way is to have one backlog for each logical sstable set, but on the other hand, backlog is also a function of compaction backlog on the tablet level. If we see backlog going very high across the logical sets (repaired and unrepaired), we might prefer giving up on incremental repair than affecting service with high latency. We could revisit this later too. 

> 
> > Also it's needed for correctness because a sstable can move quickly
> > across the logical sets, and having one tracker for each logical
> > set could cause the sstable to not be erased in the old set it
> > belonged to;
> 
> Why not? Provided we calculate atomically.

Backlog tracker is stateful. It keeps state for each sstable in the set. So we could have problems when a sstable move from one logical set to another, we subtract the sstable in the tracker of new set, and never do it in its old tracker. So we're susceptible to this race, since we kind of tie backlog tracker to a single sstables::sstable_set.

raphaelsc commented:
v2:
- rebased
- added compaction_group_view for sstables undergoing repair (repairing), this might simplify a lot the synchronization work since once a sstable is marked as repairing, naturally it will be excluded for compaction from the other groups, repairing sstables can only be compacted between themselves which is reasonable. When doing the synchronization work, we can still decide to stop compaction on a compaction_group_view level, if needed. The changes are in the last patch.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17960/) | :white_check_mark: | thor |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17960/) | :white_check_mark: | spider6.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17960/) | :x: | spider6.cloudius-systems.com |

#### Build Details:
- Duration: 21 min

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17964/) | :white_check_mark: | sif |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17964/) | :white_check_mark: | i-030a5267ad51e9f3d |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17964/) | :white_check_mark: | i-030a5267ad51e9f3d |

#### Build Details:
- Duration: 52 min

raphaelsc commented:
> #### :green_circle: CI State: SUCCESS
> 
> | Mode | Stage | Status | Node |
> |------|-------|--------|------|
> |  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17964/) | :white_check_mark: | sif |
> | release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17964/) | :white_check_mark: | i-030a5267ad51e9f3d |
> |  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17964/) | :white_check_mark: | i-030a5267ad51e9f3d |
> 
> #### Build Details:
> - Duration: 52 min
> 

How come? @yaronkaikov

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17968/) | :white_check_mark: | ran |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17968/) | :white_check_mark: | i-09a6b43277f607a5d |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17968/) | :white_check_mark: | i-09a6b43277f607a5d |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17968/)<br> :small_blue_diamond: test/boost/compaction_group_test.cc<br> :small_blue_diamond: test/boost/cql_query_large_test.cc<br> :small_blue_diamond: test/boost/database_test.cc<br> :small_blue_diamond: test/boost/incremental_compaction_test.cc<br> :small_blue_diamond: test/boost/sstable_3_x_test.cc<br> :small_blue_diamond: test/boost/sstable_compaction_test.cc<br> :small_blue_diamond: test/boost/sstable_resharding_test.cc<br> :small_blue_diamond: test/boost/sstable_test.hh<br> :small_blue_diamond: test/boost/view_complex_test.cc | :white_check_mark: | i-09a6b43277f607a5d |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17968/) | :white_check_mark: | i-00ea1e2198618adf8 |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17968/) | :white_check_mark: | i-00ea1e2198618adf8 |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17968/)<br> :small_blue_diamond: test/boost/compaction_group_test.cc<br> :small_blue_diamond: test/boost/cql_query_large_test.cc<br> :small_blue_diamond: test/boost/database_test.cc<br> :small_blue_diamond: test/boost/incremental_compaction_test.cc<br> :small_blue_diamond: test/boost/sstable_3_x_test.cc<br> :small_blue_diamond: test/boost/sstable_compaction_test.cc<br> :small_blue_diamond: test/boost/sstable_resharding_test.cc<br> :small_blue_diamond: test/boost/sstable_test.hh<br> :small_blue_diamond: test/boost/view_complex_test.cc | :white_check_mark: | i-00ea1e2198618adf8 |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17968/) | :white_check_mark: | i-044106085cba6803c |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17968/) | :white_check_mark: | i-044106085cba6803c |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17968/)<br> :small_blue_diamond: test/boost/compaction_group_test.cc<br> :small_blue_diamond: test/boost/cql_query_large_test.cc<br> :small_blue_diamond: test/boost/database_test.cc<br> :small_blue_diamond: test/boost/incremental_compaction_test.cc<br> :small_blue_diamond: test/boost/sstable_3_x_test.cc<br> :small_blue_diamond: test/boost/sstable_compaction_test.cc<br> :small_blue_diamond: test/boost/sstable_resharding_test.cc<br> :small_blue_diamond: test/boost/sstable_test.hh<br> :small_blue_diamond: test/boost/view_complex_test.cc | :white_check_mark: | i-044106085cba6803c |
| dtest | [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5757/) | :white_check_mark: |  |
|  | [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8951/) | :white_check_mark: |  |
|  | [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/17033/) | :white_check_mark: |  |

#### Build Details:
- Duration: 1 hr 37 min

asias commented on replica/compaction_group.hh:
```diff
@@ -38,8 +46,18 @@ using enable_backlog_tracker = bool_class<class enable_backlog_tracker_tag>;
+    // logical compaction group representing the unrepaired sstables.
+    std::unique_ptr<compaction_group_view> _unrepaired_view;
+    // logical compaction group representing the sstables undergoing repair.
+    std::unique_ptr<compaction_group_view> _repairing_view;
```
I think we do not need the repairing view.  We are not going to compact the sstables being repaired. The pattern like below used in multiple places will trigger compaction on the repairing sstables, which we do not really want.  

```
      for (auto view : all_views()) {
        _t._compaction_manager.submit(*view);
      }

```

raphaelsc replied:
Is it a problem that repairing sstables are compacted with one another? Cannot we only perform the synchronization when repair ends and we want to mark them as repaired? I ask this because compacting repairing sstables might be good for reads in some scenario where unrepaired sstables, selected by repair, had a high compaction backlog. If we can do this, it would be really good for the queries in such scenarios.

raphaelsc replied:
A good reason for having repairing view is that we can easily iterate through the repairing sstables whenever we want. It can also be good on educating one on how things work. Even if we want to exclude them for compaction, we could still keep it to make the existence of the state explicit. @avikivity asked for it.

asias replied:
If we should compact the repairing sstables or not are discussed previous. The decision is not to compact them at all. This is a huge simplification. This avoids compaction to mark sstables as repaired in case compaction last longer than repair. 

The gain of compacting repairing ssables is small.  Avi even argued that it is ok to serialize compaction and tablet repair all together. 

If the latency exceeds some threshold, we can always abort repair. 

We do not need to iterating repairing sstables for compaction and repair synchronization when pushing synchronization to replica/table.cc. See v4, the synchronization is done  in replica/table.cc already.

raphaelsc replied:
> If we should compact the repairing sstables or not are discussed previous. The decision is not to compact them at all. This is a huge simplification. This avoids compaction to mark sstables as repaired in case compaction last longer than repair.
> 

Actually I am suggesting to abort compaction on repair start and on repair end, so we can easily mark them with any state.

> The gain of compacting repairing ssables is small. Avi even argued that it is ok to serialize compaction and tablet repair all together.

Why is it ok? @avikivity. It's ok in normal situations when unrepaired set is a small subset. But unrepaired set can be even the total set, when transitioning from regular repair to incremental repair, or some other condition where incremental repair was delayed for whichever reason.

I think we can very easily allow compaction on repairing set (while allowing the repairing set existence to be explicit with its own compaction_group_view, as requested by @avikivity himself):

see:
```
repair start
    abort and disable compaction on unrepaired
    select all unrepaired
    mark all unrepaired as repairing
repair end
    abort and disable compaction on repairing
    select all repairing
    mark all repairing as repaired
```

It's very simple synchronization approach, while still allowing repairing sstables to be compacted with one another. I am very worried about disallowing compaction. If we can still allow compaction on repairing with a simple approach, why not? Let's also think about abnormal conditions.

asias replied:
This is from Avi from the discussion if we want to enable regular compaction when tablet repair starts. 

```Note, with 100 tablets/shard there is a 1% chance that our tablet is being compacted (actually higher since several compaction tiers are running simultaneously on a shard). So disabling compaction isn't a big deal, likely compaction isn't touching our tablet anyway.```

asias replied:
Abort  the compaction of repairing sstables after repair is done is wasteful 1) It is wasteful to drop work perhaps 99% done in general 2) when repair ends, the repairing sstables will be compacted with the larger existing unrepaired sstables anyway. 

Another argument is that if a tablet desperately needs compaction, tablet repair is not supposed to be scheduled by the tablet scheduler in the first place, we can do the compaction and then let the scheduler to schedule repair. 

Btw, indeed, we can ensure only repair marks sstables as repaired by aborting or waiting for pending repairing sstables compaction.

asias replied:
I'd like to keep it simple for the initial incremental repair. It is not a must to compact repairing sstables. I propose to make it a follow up feature if it is really needed, especially we can do the optimization with different solutions, e.g, allow repairing compaction when repair runs, do not schedule repair if repairing sstables set needs compaction, compact repairing sstables before repair.

asias replied:
> Not exactly true. Think of overwrite intensive workload with high temporal locality. Also, reducing sstable count is a big deal. Allowing sstable to run on repairing set can allow compaction to reduce sstable count and get rid of redundant data.

SStables from new writes will not in repairing set during repair period. They will not compact with repairing set whether we compact repairng sstable set or not during repair period. 

> I think we shouldn't mix repair and compaction. Compaction belongs to storage layer, repair lies in a higher level. The synchronization done in replica layer is fine, since this layer is aware of both.

I think it is not a mix. It is a scheduling decision. It is like the scheduler schedules tablet migration and tablet repair, but it does not mean we mix tablet migration and repair. Actually, it is isolating repair and compaction more because we need less synchronization, e.g., no need to abort compaction after repair is done. 

> There's nothing special with tablets that would significantly change how we perceive compaction.
Compaction is not something you schedule, control, or observe by coordinator. It's a purely replica entity. You're trying to push a mechanism to a layer it doesn't belong. Compaction lies in the storage layer of replica, and it should remain there. In the incremental repair, the synchronization remains in the replica layer and it should stay there.

It is true how we compact sstables is unchanged. However, higher level entity like the tablet scheduler can definitely make better decisions cluster wide based on the info which the compaction manager on a local node does not have. E.g., it can answer the questions, if there are no urgent compaction needed for a tablet, so that tablet repair can go ahead without compacting unrepaired sstables which will become repairing sstables (avoiding any wait or abort of existing compaction before or start repair, which avoids wasting 99% done compaction or time to delay the repair simple waiting for one replica doing compacting).  Or if it is urgent to migrate some tablets replica to a shard, so please do not trigger more compaction on this shard to allow the tablet migration to finish as soon as possible.

raphaelsc replied:
> > Not exactly true. Think of overwrite intensive workload with high temporal locality. Also, reducing sstable count is a big deal. Allowing sstable to run on repairing set can allow compaction to reduce sstable count and get rid of redundant data.
> 
> SStables from new writes will not in repairing set during repair period. They will not compact with repairing set whether we compact repairng sstable set or not during repair period. 
> 

I am talking about older writes.


> > I think we shouldn't mix repair and compaction. Compaction belongs to storage layer, repair lies in a higher level. The synchronization done in replica layer is fine, since this layer is aware of both.
> 
> I think it is not a mix. It is a scheduling decision. It is like the scheduler schedules tablet migration and tablet repair, but it does not mean we mix tablet migration and repair. Actually, it is isolating repair and compaction more because we need less synchronization, e.g., no need to abort compaction after repair is done. 
> 
> > There's nothing special with tablets that would significantly change how we perceive compaction.
> Compaction is not something you schedule, control, or observe by coordinator. It's a purely replica entity. You're trying to push a mechanism to a layer it doesn't belong. Compaction lies in the storage layer of replica, and it should remain there. In the incremental repair, the synchronization remains in the replica layer and it should stay there.
> 
> It is true how we compact sstables is unchanged. However, higher level entity like the tablet scheduler can definitely make better decisions cluster wide based on the info which the compaction manager on a local node does not have. E.g., it can answer the questions, if there are no urgent compaction needed for a tablet, so that tablet repair can go ahead without compacting unrepaired sstables which will become repairing sstables (avoiding any wait or abort of existing compaction before or start repair, which avoids wasting 99% done compaction or time to delay the repair simple waiting for one replica doing compacting).  Or if it is urgent to migrate some tablets replica to a shard, so please do not trigger more compaction on this shard to allow the tablet migration to finish as soon as possible. 

Then also push memtable flush decision. Compaction is just storage maintenance. It is part of lsm tree. It would be a mistake in my opinion.

asias commented on replica/compaction_group.hh:
```diff
@@ -173,14 +198,24 @@ public:
+    // With static sharding, i.e. vnodes, there will be only one active view.
+    compaction::compaction_group_view& as_view_for_static_sharding() const;
+    // Default view to be used on newly created sstables, e.g. those produced by repair or memtable.
+    compaction::compaction_group_view& default_view_for_new_sstables() const;
```
SStables  produced by repair itself are repaired sstable, while the sstables produced by memtable are unrepaired.  We'd better avoid the default view, it is hard to reason about what is default.  I think it is better to use repaired_view or unrepaired_view explicitly.

raphaelsc replied:
Sure, will address this.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18139/) | :white_check_mark: | godzilla |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18139/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18139/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18139/)<br> :small_blue_diamond: test/boost/compaction_group_test.cc<br> :small_blue_diamond: test/boost/cql_query_large_test.cc<br> :small_blue_diamond: test/boost/database_test.cc<br> :small_blue_diamond: test/boost/incremental_compaction_test.cc<br> :small_blue_diamond: test/boost/sstable_3_x_test.cc<br> :small_blue_diamond: test/boost/sstable_compaction_test.cc<br> :small_blue_diamond: test/boost/sstable_resharding_test.cc<br> :small_blue_diamond: test/boost/view_complex_test.cc | :white_check_mark: | spider1.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18139/) | :white_check_mark: | spider8.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18139/) | :white_check_mark: | spider8.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18139/)<br> :small_blue_diamond: test/boost/compaction_group_test.cc<br> :small_blue_diamond: test/boost/cql_query_large_test.cc<br> :small_blue_diamond: test/boost/database_test.cc<br> :small_blue_diamond: test/boost/incremental_compaction_test.cc<br> :small_blue_diamond: test/boost/sstable_3_x_test.cc<br> :small_blue_diamond: test/boost/sstable_compaction_test.cc<br> :small_blue_diamond: test/boost/sstable_resharding_test.cc<br> :small_blue_diamond: test/boost/view_complex_test.cc | :white_check_mark: | spider8.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18139/) | :white_check_mark: | spider3.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18139/) | :x: | spider3.cloudius-systems.com |
| dtest | [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5919/) | :white_check_mark: |  |
|  | [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/9146/) | :x: |  |
|  | [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/17238/) | :white_check_mark: |  |

#### Failed Tests (4/69840):
- [max_ongoing_compaction_test.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18139/testReport/junit/boost.sstable_compaction_test/cc/debug___test_py___debug___max_ongoing_compaction_test_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+max_ongoing_compaction_test.debug.1)
- [max_ongoing_compaction_test.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18139/testReport/junit/boost.sstable_compaction_test/cc/debug___test_py___debug___max_ongoing_compaction_test_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+max_ongoing_compaction_test.debug.2)
- [max_ongoing_compaction_test.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18139/testReport/junit/boost.sstable_compaction_test/cc/debug___test_py___debug___max_ongoing_compaction_test_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+max_ongoing_compaction_test.debug.3)
- [test_trigger_snapshot_transfer](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18139/testReport/junit/topology_operation_during_upgrade_test/TopologyOperationWithMixedCLusterTest/release___dtest_with_consistent_topology_changes___test_trigger_snapshot_transfer) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_trigger_snapshot_transfer)

#### Build Details:
- Duration: 6 hr 42 min

asias commented on replica/table.cc:
```diff
@@ -3961,20 +4064,27 @@ compaction_group* table::try_get_compaction_group_with_static_sharding() const {
+future<> table::parallel_foreach_compaction_group_view(std::function<future<>(compaction_group_view&)> action) {
     return parallel_foreach_compaction_group([action = std::move(action)] (compaction_group& cg) -> future<> {
-       return action(cg.as_table_state());
+       for (auto view : cg.all_compacting_views()) {
```
Here, table::parallel_foreach_compaction_group_view suggests we are walking through all the compaction group views, yet we are only walking through compacting views. 

There is no user of the repairing compaction_group_view.  It is pretty trivial to add it later when we really need it. I would recommend to drop the repairing review for now.

raphaelsc replied:
Anyway, we can keep the classification of repairing, it's just that repairing sstables will be automatically unavailable for compaction.

raphaelsc replied:
To add a little bit more, to mark unrepaired as repairing, you just need to disable / abort compaction on unrepaired view, then just mark all the sstables there as repairing. then you can enable compaction back. so newly created unrepaire can be compacted.

asias replied:
We can also make repairing sstables unavailable for compaction automatically without introducing the repairing compaction view. As long as a sstable is marked as repairing, it will be executed from the unrepaired view set. This is what we have in the current version of the incremental repair PR. 

The filter needed in this PR will exclude the repairing sstables in the unrepaired view anyway, so it is redundant to have a repairing compaction view which is not used by anyone. 

IMHO, the fact table::parallel_foreach_compaction_group_view misses one of the compaction view adds unnecessary complexity which we can avoid.   

Another thing regarding the repaired, repairing, unrepaired classification, it assumes the repairing is unrepaired, this can change. For instance, when we repairing the repaired set, the repairing sstable could be repaired sstable as well. We should not compact repairing sstables from repaired and unrepaired together.  This classification itself is not perfect.  

As for now, I really think keep repaired and unrepaired classification is more than enough.

raphaelsc replied:
> We can also make repairing sstables unavailable for compaction automatically without introducing the repairing compaction view. As long as a sstable is marked as repairing, it will be executed from the unrepaired view set. This is what we have in the current version of the incremental repair PR.
> 
> The filter needed in this PR will exclude the repairing sstables in the unrepaired view anyway, so it is redundant to have a repairing compaction view which is not used by anyone.

This is exactly what I plan to do. To keep the classification_func, with repairing as a option, but it will be excluded with the filter, since no view will exist for repairing.

> 
> IMHO, the fact table::parallel_foreach_compaction_group_view misses one of the compaction view adds unnecessary complexity which we can avoid.

Agreed.

> 
> Another thing regarding the repaired, repairing, unrepaired classification, it assumes the repairing is unrepaired, this can change. For instance, when we repairing the repaired set, the repairing sstable could be repaired sstable as well. We should not compact repairing sstables from repaired and unrepaired together. This classification itself is not perfect.
> 
> As for now, I really think keep repaired and unrepaired classification is more than enough.

The problem is that unrepaired compaction will include repairing. We can consider unrepaired and repaired as not repairing. Unrepaired or repaired that are being repaired will not actually belong to any compaction group view, since they're not eligible for compaction and actually belongs to repairing group now.

raphaelsc replied:
> If we keep the repairing view, we need to think about what we should adjust when the view_for_sstable(sst) returns the repairing view. E.g.,
> 
> ```
> auto& view = cg.view_for_sstable(sst);
> co_return co_await _t.get_compaction_manager().maybe_split_sstable(sst, view, split_compaction_options());
> ```

This happens when attaching a new sstable. Think of this like extension of sstable writing, not actually compaction. So it's fine the we split here, since this sstable is not loaded yet into sstable set.

asias commented:
btw, there is a typo in the subject s/Preparate/Prepare/

raphaelsc commented:
> btw, there is a typo in the subject s/Preparate/Prepare/

It was a confusion of mine with my mother language :-) Let me fix it.

asias commented on replica/table.cc:
```diff
@@ -3950,8 +4024,28 @@ const compaction_manager& compaction_group::get_compaction_manager() const noexc
+compaction::compaction_group_view& compaction_group::view_for_sstable(const sstables::shared_sstable& sst) const {
+    switch (_repair_sstable_classifier(sst)) {
+        case repair_sstable_classification::unrepaired: return *_unrepaired_view;
+        case repair_sstable_classification::repairing: break; // we need _repairing_view to support this.
```
This will cause the node to abort  if we pass a repairing sstable to this function.

asias replied:
We could pass a compaction_group_view, but this means we need to check on all the caller site for nullptr.

raphaelsc commented:
v4: bring back repairing view but disable compaction on it altogether.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18247/) | :white_check_mark: | ran |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18247/) | :white_check_mark: | spider6.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18247/) | :x: | spider6.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18247/) | :white_check_mark: | spider3.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18247/) | :x: | spider3.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18247/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18247/) | :x: | spider1.cloudius-systems.com |
| dtest | [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/17389/) | :x: |  |
|  | [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/6065/) | :x: |  |

#### Failed Tests (850/1136):
- [test_replay_after_schema_change](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18247/testReport/junit/batch_test/TestBatch/release___dtest_with_tablets___test_replay_after_schema_change) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replay_after_schema_change)
- [test_replay_after_schema_change](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18247/testReport/junit/batch_test/TestBatch/release___dtest_with_gossip_topology_changes___test_replay_after_schema_change) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replay_after_schema_change)
- [test_local_quorum_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18247/testReport/junit/bootstrap_test/TestBootstrap/release___dtest_with_tablets___test_local_quorum_bootstrap) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_local_quorum_bootstrap)
- [test_local_quorum_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18247/testReport/junit/bootstrap_test/TestBootstrap/release___dtest_with_gossip_topology_changes___test_local_quorum_bootstrap) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_local_quorum_bootstrap)
- [test_add_field_with_cdc[Multi_DC_cluster]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18247/testReport/junit/cdc_test/TestCdc/release___dtest_with_gossip_topology_changes___test_add_field_with_cdc_Multi_DC_cluster_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_add_field_with_cdc[Multi_DC_cluster])
- [test_add_field_with_cdc[Single_cluster]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18247/testReport/junit/cdc_test/TestCdc/release___dtest_with_gossip_topology_changes___test_add_field_with_cdc_Single_cluster_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_add_field_with_cdc[Single_cluster])
- [test_cluster_expansion_with_cdc[Multi_DC_cluster]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18247/testReport/junit/cdc_test/TestCdc/release___dtest_with_gossip_topology_changes___test_cluster_expansion_with_cdc_Multi_DC_cluster_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_expansion_with_cdc[Multi_DC_cluster])
- [test_cluster_expansion_with_cdc[Multi_DC_cluster]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18247/testReport/junit/cdc_test/TestCdc/release___dtest_with_gossip_topology_changes___test_cluster_expansion_with_cdc_Multi_DC_cluster__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_expansion_with_cdc[Multi_DC_cluster])
- [test_cluster_expansion_with_cdc[Single_cluster]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18247/testReport/junit/cdc_test/TestCdc/release___dtest_with_gossip_topology_changes___test_cluster_expansion_with_cdc_Single_cluster_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_expansion_with_cdc[Single_cluster])
- [test_cluster_expansion_with_cdc[Single_cluster]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18247/testReport/junit/cdc_test/TestCdc/release___dtest_with_gossip_topology_changes___test_cluster_expansion_with_cdc_Single_cluster__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_expansion_with_cdc[Single_cluster])

#### Build Details:
- Duration: 20 hr

raphaelsc commented:
> > v4: bring back repairing view but disable compaction on it altogether.
> 
> Did you run any test on this version? I seems there was a serious bug somewhere. I saw a single scylla node won't boot up with this version. It hangs and only kill -9 could stop the node.
> 
> ```
> INFO  2025-07-14 16:01:37,557 [shard 0:main] repair - Loading repair history for keyspace=system, table=cdc_local, table_uuid=0bcaffd4-0c83-3ead-ad13-dc1d5015b77c
> INFO  2025-07-14 16:01:37,557 [shard 0:main] repair - Loading repair history for keyspace=system_schema, table=indexes, table_uuid=0feb57ac-311f-382f-ba6d-9024d305702f
> INFO  2025-07-14 16:01:37,557 [shard 0:main] repair - Loading repair history for keyspace=system, table=raft_snapshots, table_uuid=6fa537a7-7e97-344e-9034-b69c7ef69be6
> INFO  2025-07-14 16:01:37,557 [shard 0:main] repair - Loading repair history for keyspace=system, table=group0_history, table_uuid=027e42f5-683a-3ed7-b404-a0100762063c
> INFO  2025-07-14 16:01:37,557 [shard 0:main] repair - Loading repair history for keyspace=system, table=discovery, table_uuid=1ef3d792-4f26-3706-a049-fc39024f49c5
> INFO  2025-07-14 16:01:37,557 [shard 0:main] repair - Loading repair history for keyspace=system_schema, table=scylla_aggregates, table_uuid=08d8cb28-9202-3371-a968-ad926e0fdc37
> INFO  2025-07-14 16:01:37,557 [shard 0:main] repair - Loading repair history for keyspace=system, table=view_build_status_v2, table_uuid=e0ab84ed-510d-374c-bcc0-973d28d13640
> INFO  2025-07-14 16:01:37,557 [shard 0:main] repair - Loading repair history for keyspace=system, table=dicts, table_uuid=c9c77c6f-c08c-3303-9b94-65dd3d25f3ba
> INFO  2025-07-14 16:01:37,557 [shard 0:main] repair - Loading repair history for keyspace=system, table=broadcast_kv_store, table_uuid=38b93419-1653-3153-b3e5-ea238391b2db
> INFO  2025-07-14 16:01:37,557 [shard 0:main] repair - Loading repair history for keyspace=system, table=scylla_table_schema_history, table_uuid=0191a53e-40f0-31d4-9171-b0d19ffb17b4
> INFO  2025-07-14 16:01:37,557 [shard 0:main] repair - Loading repair history for keyspace=system, table=tablets, table_uuid=fd4f7a46-96bd-3e73-91bf-99eb77e82a5c
> INFO  2025-07-14 16:01:37,563 [shard 0:main] init - starting Raft Group Registry service
> INFO  2025-07-14 16:01:37,563 [shard 0:main] init - initializing system schema
> INFO  2025-07-14 16:01:37,563 [shard 0:main] legacy_schema_migrator - Moving 0 keyspaces from legacy schema tables to the new schema keyspace (system_schema)
> INFO  2025-07-14 16:01:37,567 [shard 0:main] legacy_schema_migrator - Dropping legacy schema tables
> INFO  2025-07-14 16:01:37,567 [shard 0:main] database - Dropping system.schema_aggregates without snapshot
> INFO  2025-07-14 16:01:37,567 [shard 0:main] database - Truncating system.schema_aggregates without snapshot
> INFO  2025-07-14 16:01:37,567 [shard 0:main] database - Dropping system.schema_functions without snapshot
> INFO  2025-07-14 16:01:37,567 [shard 0:main] database - Truncating system.schema_functions without snapshot
> INFO  2025-07-14 16:01:37,567 [shard 0:main] database - Dropping system.schema_usertypes without snapshot
> INFO  2025-07-14 16:01:37,567 [shard 0:main] database - Truncating system.schema_usertypes without snapshot
> INFO  2025-07-14 16:01:37,567 [shard 0:main] database - Dropping system.schema_columns without snapshot
> INFO  2025-07-14 16:01:37,567 [shard 0:main] database - Truncating system.schema_columns without snapshot
> INFO  2025-07-14 16:01:37,567 [shard 0:main] database - Dropping system.schema_triggers without snapshot
> INFO  2025-07-14 16:01:37,567 [shard 0:main] database - Truncating system.schema_triggers without snapshot
> INFO  2025-07-14 16:01:37,567 [shard 0:main] database - Dropping system.schema_columnfamilies without snapshot
> INFO  2025-07-14 16:01:37,567 [shard 0:main] database - Truncating system.schema_columnfamilies without snapshot
> INFO  2025-07-14 16:01:37,567 [shard 0:main] database - Dropping system.schema_keyspaces without snapshot
> INFO  2025-07-14 16:01:37,567 [shard 0:main] database - Truncating system.schema_keyspaces without snapshot
> INFO  2025-07-14 16:01:37,667 [shard 0:main] sstable - SSTable /home/asias/src/cloudius-systems/scylla/mycluster/tmp/1/data/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/me-1-big-Data.db has numerical generation. SSTable identifier in scylla_metadata set to c47a8d30-6088-11f0-9d14-a2a2d03d0497
> INFO  2025-07-14 16:01:37,701 [shard 0:main] database - Truncated system.schema_keyspaces
> INFO  2025-07-14 16:01:37,701 [shard 0:main] database - Truncated system.schema_columnfamilies
> INFO  2025-07-14 16:01:37,701 [shard 0:main] database - Truncated system.schema_triggers
> INFO  2025-07-14 16:01:37,701 [shard 0:main] database - Truncated system.schema_columns
> INFO  2025-07-14 16:01:37,701 [shard 0:main] database - Truncated system.schema_usertypes
> INFO  2025-07-14 16:01:37,701 [shard 0:main] database - Truncated system.schema_functions
> INFO  2025-07-14 16:01:37,701 [shard 0:main] database - Truncated system.schema_aggregates
> INFO  2025-07-14 16:01:46,418 [shard 0:main] sstable - SSTable /home/asias/src/cloudius-systems/scylla/mycluster/tmp/1/data/system/local-7ad54392bcdd35a684174e047860b377/me-1-big-Data.db has numerical generation. SSTable identifier in scylla_metadata set to c9b1d920-6088-11f0-9d14-a2a2d03d0497
> ```

I think it's fixed now. Please pull latest version of the branch and retry

asias commented:
@raphaelsc Thanks. I can confirm the hang is gone with the fix. Can you take a look at the test max_ongoing_compaction_test error. I saw it as well here https://github.com/scylladb/scylladb/pull/24291#issuecomment-2965651093.

bhalevy commented:
@raphaelsc https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18348/testReport/boost.sstable_compaction_test/cc/debug___test_py___debug___max_ongoing_compaction_test_debug_1/ fails consistently in debug, please reproduce and fix.

raphaelsc commented:
> > > @raphaelsc [jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18348/testReport/boost.sstable_compaction_test/cc/debug___test_py___debug___max_ongoing_compaction_test_debug_1](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18348/testReport/boost.sstable_compaction_test/cc/debug___test_py___debug___max_ongoing_compaction_test_debug_1/) fails consistently in debug, please reproduce and fix.
> > 
> > 
> > Elsewhere?
> 
> I don't understand the question

If you reproduced this problem in another branch that doesn't contain these changes

denesb commented on compaction/compaction_group_view.hh:
```diff
@@ -38,8 +38,8 @@ public:
-    virtual const sstables::sstable_set& main_sstable_set() const = 0;
-    virtual const sstables::sstable_set& maintenance_sstable_set() const = 0;
+    virtual future<lw_shared_ptr<const sstables::sstable_set>> main_sstable_set() const = 0;
+    virtual future<lw_shared_ptr<const sstables::sstable_set>> maintenance_sstable_set() const = 0;
```
Do these futurizations (this and the previous one) create opening for compaction races? E.g. one compaction fiber is retrieving candidates, another one is started and manages to start. Do we have locks to prevent such scenarios?

denesb commented on replica/compaction_group.hh:
```diff
@@ -38,8 +48,20 @@ using enable_backlog_tracker = bool_class<class enable_backlog_tracker_tag>;
     class compaction_group_view;
-    std::unique_ptr<compaction_group_view> _compaction_group_view;
+    // This is held throughout group lifetime, in order to have compaction disabled on non-compacting views.
+    std::vector<compaction_manager::compaction_reenabler> _compaction_disabler_for_views;
```
Can't we permanently disable compaction, without having to keep some RAII object alive?

raphaelsc replied:
We could, but I felt like I could reuse something existing, rather than leaking a count here https://github.com/scylladb/scylladb/blob/master/compaction/compaction_state.hh#L31

What do you think? Maybe I can just set it to 1 in add_with_compaction_disabled() and that's it. And document that a compaction_group_view added through that interface in manager will never have compaciton enabled. The reason we want to register non-compacting view is for compatibility with table code iterating on all views and expecting them all to be registered in the manager.

denesb replied:
Ok, we can keep it as-is.

denesb reviewed (commented):
LGTM (weak approve)
I have some questions, mostly due to my weak understanding of the this code.

avikivity commented on compaction/compaction_group_view.hh:
```diff
@@ -38,8 +38,8 @@ public:
     // min threshold as defined by table.
     virtual unsigned min_compaction_threshold() const noexcept = 0;
     virtual bool compaction_enforce_min_threshold() const noexcept = 0;
```
Maybe comments too.

raphaelsc replied:
Sure.

raphaelsc commented on replica/compaction_group.hh:
```diff
@@ -38,8 +48,20 @@ using enable_backlog_tracker = bool_class<class enable_backlog_tracker_tag>;
+    // together.
     class compaction_group_view;
-    std::unique_ptr<compaction_group_view> _compaction_group_view;
+    // This is held throughout group lifetime, in order to have compaction disabled on non-compacting views.
```
> It's odd that a compaction group has multiple views. Why not have multiple compaction groups?

That was our first consideration.

but then with tablets, you will have unrepaired, repairing and repaired groups **for each of** main, left, right, merging.

The idea with view is that it's a logical compaction group, while replica::compaction_group is actually a compaction_group_set. Where each logical group will share the same instance of sstable set. Maybe we should rename replica::compaction_group to replica::compaction_group_set? and rename compaction_group_view to just compaction_group?

Another good reason for that, is that we don't have the complexity of moving sstables across groups (sstable sets) as soon as they change state. So the single source of truth is group0 which determines the state of every sstable based on its persisted repaired_at.

asias replied:
I'd suggest not to add more renaming for now. Let's do it after this series is consumed by incremental repair.

avikivity replied:
Please add a test that populates a compaction group with fake sstables, shows they are compacted together, then populates the compaction group with the same fake sstables but different repaired_at settings, and show that compaction doesn't cross repair boundaries.

It's important to have those things tested close to the implementation and not only as part of the entire system.

raphaelsc replied:
> Please add a test that populates a compaction group with fake sstables, shows they are compacted together, then populates the compaction group with the same fake sstables but different repaired_at settings, and show that compaction doesn't cross repair boundaries.
> 
> It's important to have those things tested close to the implementation and not only as part of the entire system.

I will contribute this test on stop of @asias series since this repaired_at setting and its handling only comes with his series.

asias commented:
> [dev / test.py - dev / cluster.tasks.test_tablet_tasks.test_tablet_resize_revoked.dev.1](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18631/testReport/junit/cluster.tasks/test_tablet_tasks/dev___test_py___dev___test_tablet_resize_revoked_dev_1/)

I saw this error too. The following fixed the test for me. 

```
commit c98f99812fd52d0b598a00efe0bc4e207b668e91 (HEAD -> incremental_repair, asias-pub/incremental_repair)
Author: Asias He <asias@scylladb.com>
Date:   Tue Jul 29 08:39:18 2025 +0800

    test.py: Adjust test_tablet_resize_revoked for incremental repair
    
    With incremental repair, we have doubled the tasks for repaired and
    unrepaired set. Increase the task number in the check.

diff --git a/test/cluster/tasks/test_tablet_tasks.py b/test/cluster/tasks/test_tablet_tasks.py
index 98e6299552d..0626124c3b3 100644
--- a/test/cluster/tasks/test_tablet_tasks.py
+++ b/test/cluster/tasks/test_tablet_tasks.py
@@ -525,7 +525,8 @@ async def test_tablet_resize_revoked(manager: ManagerClient):
 
         async def wait_for_task(task_id):
             status = await tm.wait_for_task(servers[0].ip_addr, task_id)
-            check_task_status(status, ["suspended"], "split", "table", False, keyspace, table1, [0, 1, 2])
+            # With incremental repair, we have doubled the tasks for repaired and unrepaired set
+            check_task_status(status, ["suspended"], "split", "table", False, keyspace, table1, [0, 1, 2, 3, 4, 5])
 
         await asyncio.gather(revoke_resize(log, mark), wait_for_task(task0.task_id))
```

avikivity commented on test/boost/compaction_group_test.cc:
```diff
@@ -202,3 +202,70 @@ SEASTAR_TEST_CASE(basic_compaction_group_splitting_test) {
+        t.set_repair_sstable_classifier(repair_sstable_classifier);
+
+        for (int i = 0; i < 4; i++) {
+            t->compact_all_sstables({}).get();
```
This is like major compaction?

raphaelsc replied:
> This is like major compaction?

this is what major compaction calls. stresses the same path as minor / regular for segregating sstables into the logical groups.

denesb commented:
CI is dead.

asias commented:
> #### 🔴 CI State: FAILURE
> Mode	Stage	Status	Node
> [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18692/)	✅	thor
> dev	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18692/)	✅	spider2.cloudius-systems.com
> [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18692/)	✅	spider2.cloudius-systems.com
> [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18692/)
> 🔹 test/boost/compaction_group_test.cc
> 🔹 test/boost/cql_query_large_test.cc
> 🔹 test/boost/database_test.cc
> 🔹 test/boost/incremental_compaction_test.cc
> 🔹 test/boost/sstable_3_x_test.cc
> 🔹 test/boost/sstable_compaction_test.cc
> 🔹 test/boost/sstable_resharding_test.cc
> 🔹 test/boost/view_complex_test.cc
> 🔹 cluster/tasks/test_tablet_tasks	❌	spider2.cloudius-systems.com
> release	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18692/)	✅	i-0e84adb54a7400300
> [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18692/)	✅	i-0e84adb54a7400300
> [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18692/)
> 🔹 test/boost/compaction_group_test.cc
> 🔹 test/boost/cql_query_large_test.cc
> 🔹 test/boost/database_test.cc
> 🔹 test/boost/incremental_compaction_test.cc
> 🔹 test/boost/sstable_3_x_test.cc
> 🔹 test/boost/sstable_compaction_test.cc
> 🔹 test/boost/sstable_resharding_test.cc
> 🔹 test/boost/view_complex_test.cc
> 🔹 cluster/tasks/test_tablet_tasks	✅	i-0e84adb54a7400300
> debug	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18692/)	✅	i-0270ac1d17fb5e78f
> [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18692/)	❌	i-0270ac1d17fb5e78f
> dtest	[dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/6420/)	✅	
> [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/9670/)	✅	
> [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/17766/)	✅	
> #### Failed Tests (2/135150):
> * [test_tablet_resize_list.dev.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18692/testReport/junit/cluster.tasks/test_tablet_tasks/dev___Unit_Tests_Custom___dev___test_tablet_resize_list_dev_4) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_resize_list.dev.4)
> * [test_tablet_resize_list.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18692/testReport/junit/cluster.tasks/test_tablet_tasks/debug___test_py___debug___test_tablet_resize_list_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_resize_list.debug.2)
> 
> #### Build Details:
> * Duration: 13 hr

This patch should help the failed test. 

https://github.com/scylladb/scylladb/pull/24291/commits/b0fed4dfe8c152a849686d086ec36090d4817e68

```
commit b0fed4dfe8c152a849686d086ec36090d4817e68 (HEAD -> incremental_repair, asias-pub/incremental_repair)
Author: Asias He <asias@scylladb.com>
Date:   Thu Jul 31 07:44:54 2025 +0800

    test.py: Adjust test_tablet_resize_list for incremental repair
    
    With incremental repair, we have doubled the tasks for repaired and
    unrepaired set. Increase the task number in the check.

diff --git a/test/cluster/tasks/test_tablet_tasks.py b/test/cluster/tasks/test_tablet_tasks.py
index 0626124c3b3..4596120ba33 100644
--- a/test/cluster/tasks/test_tablet_tasks.py
+++ b/test/cluster/tasks/test_tablet_tasks.py
@@ -482,7 +482,9 @@ async def test_tablet_resize_list(manager: ManagerClient):
 
         status1 = await tm.get_task_status(servers[1].ip_addr, task0.task_id)
         status0 = await tm.get_task_status(servers[0].ip_addr, task0.task_id)
-        assert len(status0.children_ids) == 2
+        children_ids_len = len(status0.children_ids)
+        # With incremental repair, we have doubled the tasks for repaired and unrepaired set
+        assert children_ids in [2, 4]
         assert status0.children_ids == status1.children_ids
 
         await disable_injection(manager, servers, injection)
```

denesb commented:
Replaced by https://github.com/scylladb/scylladb/pull/25287.
